recursion-schemes.cabal

```diff
@@ -41,6 +41,9 @@ library
     StandaloneDeriving
     UndecidableInstances
 
+  include-dirs: include
+  -- includes: recursion-schemes-common.h
```

Comment:
Why is this commented out?

Reply:
It fails with it on my machine, I have no idea yet why.

Reply:
Hm, it looks like `cabal configure` is preprocessing `recursion-schemes-common.h` but doesn't have access to the definitions of `MIN_VERSION_base(4,9,0)` and friends? At least, that's what `cabal configure -v3` is giving me:

```
./include/recursion-schemes-common.h:5:21: error: missing binary operator
before token "("
#if MIN_VERSION_base(4,9,0)
^
./include/recursion-schemes-common.h:8:31: error: missing binary operator
before token "("
#elif MIN_VERSION_transformers(0,5,0)
^
./include/recursion-schemes-common.h:11:74: error: missing binary operator
before token "("
#elif MIN_VERSION_transformers_compat(0,5,0) &&
!MIN_VERSION_transformers(0,4,0)
^
```

I too don't know how to fix this, nor why `free-common.h` in the `free` package apparently doesn't suffer from this problem.

Reply:
the comparison with `free` is what I struggle indeed.

Reply:
I stared at this for a while, and I think I have a better understand of what is going on.

Here is [`free-common.h`](https://github.com/ekmett/free/blob/74947c43d0e42702e103811e0ad29778794af375/include/free-common.h#L1-L19):

```c
#ifndef MIN_VERSION_base
#define MIN_VERSION_base(x,y,z) 1
#endif

#ifndef MIN_VERSION_mtl
#define MIN_VERSION_mtl(x,y,z) 1
#endif

#if MIN_VERSION_base(4,9,0)
#define LIFTED_FUNCTOR_CLASSES 1
#else
#if MIN_VERSION_transformers(0,5,0)
#define LIFTED_FUNCTOR_CLASSES 1
#else
#if MIN_VERSION_transformers_compat(0,5,0) && !MIN_VERSION_transformers(0,4,0)
#define LIFTED_FUNCTOR_CLASSES 1
#endif
#endif
#endif
```

And for comparison, here is [`recursion-schemes-common.h`](https://github.com/phadej/recursion-schemes/blob/8f338da57f0499b7a3d8c77ebb179da2671a40b0/include/recursion-schemes-common.h#L1-L18):

```c
#ifndef MIN_VERSION_transformers_compat
#define MIN_VERSION_transformers_compat(x,y,z) 0
#endif

#if MIN_VERSION_base(4,9,0)
#define LIFTED_FUNCTOR_CLASSES 1

#elif MIN_VERSION_transformers(0,5,0)
#define LIFTED_FUNCTOR_CLASSES 1

#elif MIN_VERSION_transformers_compat(0,5,0) && !MIN_VERSION_transformers(0,4,0)
#define LIFTED_FUNCTOR_CLASSES 1
#endif

#define HAS_GENERIC (__GLASGOW_HASKELL__ >= 702)
#define HAS_GENERIC1 (__GLASGOW_HASKELL__ >= 706)

#define HAS_POLY_TYPEABLE MIN_VERSION_base(4,7,0)
```

One thing that I didn't realize until now was then when `cabal configure` runs, _it hasn't created `cabal_macros.h` yet_. In other words, no `MIN_VERSION_...` macro has been defined yet. So why does `free`'s version work, but not `recursion-schemes`'?

An important detail is that `free-common.h` explicitly defines `MIN_VERSION_base`:

```c
#ifndef MIN_VERSION_base
#define MIN_VERSION_base(x,y,z) 1
#endif
```

Moreover, take a closer look at how the body of `free-common.h` is structured:

```c
#if MIN_VERSION_base(4,9,0)
...
#else
...
#endif
```

The outermost CPP uses a single `#if`, with one conditional, so if that conditional succeeds, then the stuff within the ellipses won't be preprocessed (at least, that's my understanding). On the other hand, look at how the body of `recursion-schemes.h` is structured:

```c
#if MIN_VERSION_base(4,9,0)
...
#elif MIN_VERSION_transformers(0,5,0)
...
#elif MIN_VERSION_transformers_compat(0,5,0) && !MIN_VERSION_transformers(0,4,0)
...
#endif
```

Instead of using a single conditional, this uses _three_ conditionals, each of them depending on a different `MIN_VERSION` variant. If even a single one of these are undefined, the preprocessor will fail. Therefore, in order to fix `recursion-schemes-common.h`, we'll need to either:

1. Define all of `MIN_VERSION_base`, `MIN_VERSION_transformers`, and `MIN_VERSION_transformers_compat` beforehand, similarly to how `free-common.h` defines `MIN_VERSION_base`.
2. Restructure the body of `recursion-schemes-common.h` to more closely resemble that of `free-common.h`.

Reply:
Oh, great. I'll change this header to work based on these observations. Thanks Ryan!

Reply:
Hi @phadej , do you still plan to make those changes? There's no rush, I just want to make sure this PR is still open because somebody (plans to) complete it, not because it has been abandoned. Thanks!